[ZEPPELIN-6284] Added Flink 1.20 interpreter.

Neuw84 opened the pull request:
### What is this PR for?
A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - https://zeppelin.apache.org/contribution/contributions.html

Add Flink 1.20 support for Apache Flink interpreter. Being the latest 1.X release and LTS one it is important to be able to support it. For example we can use latest Iceberg versions with it. 

### What type of PR is it?
Feature

### Todos

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/ZEPPELIN/
* Put link here, and add [ZEPPELIN-*Jira number*] in PR title, eg. [ZEPPELIN-533]

https://issues.apache.org/jira/browse/ZEPPELIN-6284


### How should this be tested?
* Strongly recommended: add automated unit tests for any new or changed behavior
* Outline any manual steps to test the PR here.

### Screenshots (if appropriate)

### Questions:
* Does the license files need to update? No 
* Is there breaking changes for older versions? No
* Does this needs documentation? No

Commit:
[ZEPPELIN-6284] Added Flink 1.20 interpreter.

Commit:
[ZEPPELIN-6284] Added Flink 1.20 interpreter.

Neuw84 commented:
Based on previous work from 1.18 from these pull requests. 

https://github.com/apache/zeppelin/pull/4864
https://github.com/apache/zeppelin/pull/4739

jongyoul commented:
@Neuw84 Thank you for your contribution. By the way, flink test is failing now. Could you please check and fix them?

Commit:
Merge remote-tracking branch 'origin/]' into ]

# Conflicts:
#	flink/flink1.20-shims/src/main/java/org/apache/zeppelin/flink/Flink120Shims.java
#	flink/flink1.20-shims/src/main/java/org/apache/zeppelin/flink/shims120/CollectStreamTableSink.java

Neuw84 committed:
Merge branch 'apache:master' into ]

Neuw84 commented:
@jongyoul I have fixed Copilot comments and added the corresponding test file.

Neuw84 commented:
Unmmm, looking at the docs I see this weird error. Seems like it is not being able to launch the cluster, but a little bit lost here. 

```
 WARN [2025-08-21 11:04:10,021] ({flink-metrics-pekko.actor.supervisor-dispatcher-5} SupervisorActor.java[rpcActorFailed]:144) - RpcActor pekko://flink-metrics/user/rpc/MetricQueryService has failed. Shutting it down now.
java.lang.NullPointerException
	at org.slf4j.impl.Reload4jMDCAdapter.setContextMap(Reload4jMDCAdapter.java:81)
	at org.slf4j.MDC.setContextMap(MDC.java:264)
	at org.apache.flink.util.MdcUtils.lambda$withContext$0(MdcUtils.java:48)
	at org.apache.flink.runtime.rpc.pekko.PekkoRpcActor.handleControlMessage(PekkoRpcActor.java:208)
	at org.apache.pekko.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:33)
	at org.apache.pekko.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:29)
	at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
	at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
	at org.apache.pekko.japi.pf.UnitCaseStatement.applyOrElse(CaseStatements.scala:29)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
	at org.apache.pekko.actor.Actor.aroundReceive(Actor.scala:547)
	at org.apache.pekko.actor.Actor.aroundReceive$(Actor.scala:545)
	at org.apache.pekko.actor.AbstractActor.aroundReceive(AbstractActor.scala:229)
	at org.apache.pekko.actor.ActorCell.receiveMessage(ActorCell.scala:590)
	at org.apache.pekko.actor.ActorCell.invoke(ActorCell.scala:557)
	at org.apache.pekko.dispatch.Mailbox.processMailbox(Mailbox.scala:280)
	at org.apache.pekko.dispatch.Mailbox.run(Mailbox.scala:241)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
 WARN [2025-08-21 11:04:10,031] ({main} HadoopUtils.java[getHadoopConfiguration]:139) - Could not find Hadoop configuration via any of the supported methods (Flink configuration, environment variables).

```

Any help is appreciated as the Exception do not provide much info....

Indeed the code that is present here is working at least for Scala and SQL interfaces ( checked with some Notebooks) with Java 11.

Commit:
Upgraded to 1.20.2.

Hopefully this will fix the NPE with log4j.

Neuw84 commented:
After some research I wasn´t able to identify the offending library. Spend some time time trying to fix scala versions without luck on maven. 

The code works well as said on my previous comment, should be a thing of scala-reflect/compiler used on the tests but no clue.

Reamer reviewed (changes requested):
We are now using a newer JDK version and should also be using the new Java tools.

Commit:
Fixed code to use modern conventions per @Reamer suggestions.

Neuw84 commented on flink/flink1.20-shims/src/main/java/org/apache/zeppelin/flink/shims120/CollectStreamTableSink.java:
```diff
@@ -0,0 +1,97 @@
+  public CollectStreamTableSink(InetAddress targetAddress,
+                                int targetPort,
+                                TypeSerializer<Tuple2<Boolean, Row>> serializer) {
+    LOGGER.info("Use address: " + targetAddress.getHostAddress() + ":" + targetPort);
```
fixed.

Neuw84 commented on flink/flink1.20-shims/src/main/java/org/apache/zeppelin/flink/Flink120Shims.java:
```diff
@@ -0,0 +1,408 @@
+    String mode = properties.getProperty("flink.execution.mode");
+    if ("yarn-application".equalsIgnoreCase(mode)) {
+      // for yarn application mode, FLINK_HOME is container working directory
+      String flinkHome = new File(".").getAbsolutePath();
```
changed on latest commit.

Commit:
Changed how the environment is created to a modern Flink way.

Avoid Scala reflection issues on testing.

Reamer commented on flink/flink1.20-shims/src/main/java/org/apache/zeppelin/flink/Flink120Shims.java:
```diff
@@ -0,0 +1,423 @@
+    } else {
+      String flinkHome = System.getenv("FLINK_HOME");
+      if (flinkHome != null && !flinkHome.isBlank()) {
+        pyDir = Paths.get(flinkHome).resolve("opt").resolve("python");
```
Why do you use Paths.resolve() in various places?
I would have shortened it to Paths.get(“flinkHome”, “opt”, "python").

Neuw84 replied:
fixed.

Commit:
Changed how the environment is created to a modern Flink way.

Refactoring of the code.

Neuw84 commented:
After digging and digging with scala libraries I think that the error is not solvable.... I don't see any scala library collisions and tried forcing many versions. 

Seems that they changed how you need to create the environments and that should updated on ScalaInterpreter. 

We should have a codepath for 1.17 and below and one for 1.20 now.

Commit:
Fixing Protobuf version to avoid

File "/usr/share/miniconda/envs/python_3_with_flink/lib/python3.9/site-packages/google/protobuf/descriptor.py", line 796, in __new__
                                     _message.Message._CheckCalledFromGeneratedFile()
                                 TypeError: Descriptors cannot not be created directly.
                                 If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
                                 If you cannot immediately regenerate your protos, some other possible workarounds are:
                                  1. Downgrade the protobuf package to 3.20.x or lower.
                                  2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

Commit:
Fixed flink version detection code.

pan3793 commented:
FLINK-36327 replaces the `@deprecated(since = "1.18.0")` with Java's `@Deprecated`, you may want to try Flink 2.0

pan3793 commented:
> I dont know if you have noticed or not, but the error isn't actually fatal. If you try to run the cell a second time it surprisingly will work without any problems.
> 
> ```
> error reading Scala signature of org.apache.flink.table.api.ImplicitExpressionConversions: unsafe symbol x$2 (child of value <local api>) in runtime reflection universe
> ```

I also noticed that, but have no idea why, due to my limited knowledge of scala internal :(

pan3793 commented:
I think there are two directions now:

1. try Flink 2.0 to see if FLINK-36327 helps
2. eliminate usage of Flink's Scala API since it was deprecated by Flink for a long time.

Neuw84 commented:
The problem is that Flink 2.0 deletes all scala APIs, that would mean that we need to rewrite the Flink interpreter and see what are the ideal options there. SQL interface ok, but then... python+java repl? It would need a huge rearchitecture....

I will try to spend more time on this (Flink 1.20) but I can't promise anything, any help would be appreciated. As it is know it works but I would like this to be merged to have at least LTS support before migrate to Flink 2.0...

jongyoul commented:
Hello @Neuw84, I missed this PR and submit and merge my one first. I'm sorry for interrupting it. By the way, I think my PR meets the minimum bar to work with new version of Flink 1.20 but I hope it could be improved more and more. Could you please help give your experience to improve the featue? I will review it with priority. 🙏
